File: app/src/main/java/igrek/songbook/songpreview/LyricsManager.java
package igrek.songbook.songpreview;

import android.graphics.Paint;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.model.lyrics.LyricsModel;
import igrek.songbook.model.lyrics.LyricsParser;
import igrek.songbook.songpreview.autoscroll.AutoscrollService;
import igrek.songbook.songpreview.theme.LyricsThemeService;
import igrek.songbook.songpreview.transpose.ChordsTransposerManager;
import igrek.songbook.system.WindowManagerService;

public class LyricsManager {
	
	@Inject
	Lazy<ChordsTransposerManager> chordsTransposerManager;
	@Inject
	Lazy<AutoscrollService> autoscrollService;
	@Inject
	LyricsThemeService lyricsThemeService;
	@Inject
	WindowManagerService windowManagerService;
	
	private LyricsParser lyricsParser;
	private LyricsModel lyricsModel;
	private int screenW = 0;
	private Paint paint;
	private String originalFileContent;
	
	public LyricsManager() {
		DaggerIoc.getFactoryComponent().inject(this);
		lyricsParser = new LyricsParser();
	}
	
	private void reset() {
		chordsTransposerManager.get().reset();
		autoscrollService.get().reset();
	}
	
	private String normalizeContent(String content) {
		content = content.replace("\r", "");
		content = content.replace("\t", " ");
		content = content.replace("\u00A0", " "); // NO-BREAK SPACE (0xC2 0xA0)
		return content;
	}
	
	public void load(String fileContent, Integer screenW, Integer screenH, Paint paint) {
		reset();
		originalFileContent = normalizeContent(fileContent);
		
		if (screenW != null)
			this.screenW = screenW;
		if (paint != null)
			this.paint = paint;
		
		lyricsParser = new LyricsParser();
		
		parseAndTranspose(originalFileContent);
	}
	
	public void onPreviewSizeChange(int screenW, int screenH, Paint paint) {
		this.screenW = screenW;
		this.paint = paint;
		reparse();
	}
	
	public void reparse() {
		parseAndTranspose(originalFileContent);
	}
	
	public LyricsModel getCRDModel() {
		return lyricsModel;
	}
	
	private void parseAndTranspose(String originalFileContent) {
		String transposedContent = chordsTransposerManager.get()
				.transposeContent(originalFileContent);
		float realFontsize = windowManagerService.dp2px(lyricsThemeService.getFontsize());
		lyricsModel = lyricsParser.parseFileContent(transposedContent, screenW, realFontsize, paint);
	}
	
}

File: app/src/main/java/igrek/songbook/songpreview/LyricsManager.kt
package igrek.songbook.songpreview

import android.graphics.Paint
import dagger.Lazy
import igrek.songbook.dagger.DaggerIoc
import igrek.songbook.model.lyrics.LyricsModel
import igrek.songbook.model.lyrics.LyricsParser
import igrek.songbook.songpreview.autoscroll.AutoscrollService
import igrek.songbook.songpreview.theme.LyricsThemeService
import igrek.songbook.songpreview.transpose.ChordsTransposerManager
import igrek.songbook.system.WindowManagerService
import javax.inject.Inject

class LyricsManager {

    @Inject
    lateinit var chordsTransposerManager: Lazy<ChordsTransposerManager>
    @Inject
    lateinit var autoscrollService: Lazy<AutoscrollService>
    @Inject
    lateinit var lyricsThemeService: LyricsThemeService
    @Inject
    lateinit var windowManagerService: WindowManagerService

    private var lyricsParser: LyricsParser? = null
    var crdModel: LyricsModel? = null
        private set
    private var screenW = 0
    private var paint: Paint? = null
    private var originalFileContent: String? = null

    init {
        DaggerIoc.getFactoryComponent().inject(this)
    }

    private fun reset() {
        chordsTransposerManager.get().reset()
        autoscrollService.get().reset()
    }

    private fun normalizeContent(content: String): String {
        return content
                .replace("\r", "")
                .replace("\t", " ")
                .replace("\u00A0", " ") // NO-BREAK SPACE (0xC2 0xA0)
    }

    fun load(fileContent: String, screenW: Int?, screenH: Int?, paint: Paint?) {
        reset()
        originalFileContent = normalizeContent(fileContent)

        if (screenW != null)
            this.screenW = screenW
        if (paint != null)
            this.paint = paint

        val typeface = lyricsThemeService.fontTypeface!!.typeface
        lyricsParser = LyricsParser(typeface)

        parseAndTranspose(originalFileContent)
    }

    fun onPreviewSizeChange(screenW: Int, screenH: Int, paint: Paint) {
        this.screenW = screenW
        this.paint = paint
        reparse()
    }

    fun reparse() {
        parseAndTranspose(originalFileContent)
    }

    private fun parseAndTranspose(originalFileContent: String?) {
        val transposedContent = chordsTransposerManager.get()
                .transposeContent(originalFileContent)
        val realFontsize = windowManagerService.dp2px(lyricsThemeService.fontsize)
        crdModel = lyricsParser?.parseFileContent(transposedContent, screenW.toFloat(), realFontsize, paint!!)
    }

}
